Performance of Constant string printing [ci: last-only]

som-snytt opened the pull request:
Use a StringBuilder to build a String. Since quotes are always added, it's not obvious that string add is worthwhile after chars have been copied to the builder. Could also prescan for an escapable char.

Fixes https://github.com/scala/scala-dev/issues/878

som-snytt committed:
Build constant string

SethTisue commented:
Jason, want to test this out on the customer codebase...?

som-snytt commented:
OMG I thought this was just something that annoyed retronym.

retronym commented on src/reflect/scala/reflect/internal/Constants.scala:
```diff
@@ -233,23 +233,43 @@ trait Constants extends api.Constants {
+          if (i < max) loop(i + 1, escapedChar(b, text.charAt(i)) || modified)
+          else modified
+        val _ = loop(i = 0, modified = false) // could discard buffer and let intrinsic "\"" + text + "\""
+        b.append("\"").toString
```
Looks like this evolved to no longer need/use `modified`, let's remove it.

som-snytt replied:
I was about to, but I hoped it was needed.

retronym commented on src/reflect/scala/reflect/internal/Constants.scala:
```diff
@@ -233,23 +233,43 @@ trait Constants extends api.Constants {
+        val b = new StringBuilder(max).append("\"")
+        def loop(i: Int, modified: Boolean): Boolean =
+          if (i < max) loop(i + 1, escapedChar(b, text.charAt(i)) || modified)
+          else modified
```
This is a little cleaner and should be plenty fast. (`String#chars` returns an `IntStream` so no boxing to consume the values.

```
text.chars.forEach( (i: Int) => escapedChar(i.toChar))
```

som-snytt replied:
Now that we don't need modified. I will miss modified, it made everything more interesting.

som-snytt replied:
chars is jdk 9

retronym commented:
Some context -- this method is called in some cases from Zinc's Extract API phase to generate its representation of annotations.

som-snytt commented:
I may not get around to grokking the context. Presumably someone uses lots of annotations. On Dotty, the code was copied twice for "printing", so I'm not sure yet if copy/paste is called for.

SethTisue reviewed (approved):
LGTM, but I'm not sure whether to merge without Jason's approval

som-snytt commented:
I'd like @retronym guess about the last commit and comment. I didn't write a test to estimate what ratio of escapes would make it less desirable, but maybe annotations have fewer of them.

The other choice is prescan for escapable chars; then only strings with escapes must suffer construction costs. Maybe I'll add that while we're waiting.

som-snytt commented:
Actually I saw retronym doing thyme timings at https://github.com/scala/scala/commit/c3fcd4e5b0c8f03862ecf4f92fd43cf2fb8f5066 and felt bad for being lazy, so I will try to measure if this is useful.

I was at that commit because while exercising `-Wnonunit-statement`, I noticed the use of `StringBuffer` instead of `StringBuilder`. It was like a blast from the past.

lrytz commented on src/reflect/scala/reflect/internal/Constants.scala:
```diff
@@ -233,23 +233,46 @@ trait Constants extends api.Constants {
+            return 6
+          case _    => return 1
+        }
+        b.replace(i, i + 1, replace)
```
this shifts the tail of the array on each escaped char, is that more efficent than adding chars / their escapes sequentially?

som-snytt replied:
sorry I was too lazy to re-draft while I benchmark it. (At least to understand under what sort of data it is better to copy that append by char.) (I did spend a day with sbt-jmh and updated the examples to Scala 3. That took me a while, surprisingly. But I didn't get back to this exercise yet.)

lrytz commented on src/reflect/scala/reflect/internal/Constants.scala:
```diff
@@ -233,23 +233,46 @@ trait Constants extends api.Constants {
+        var i = 1 // don't escape the bookend quotes
+        while (i < b.length - 1)
+          i += escapedChar(b, i)
+        if (b.length == quoted.length) quoted else b.toString
```
Is the common case that there's nothing to escape? Currently, this allocates the result string and an unused StringBuilder initialized with the same result.

som-snytt replied:
Yes, I will benchmark pre-scan as well. Without looking, I don't remember why I didn't do that first. (Edit: I think after fiddling with code without a benchmark, I was lazy about further fiddling without data.)

lrytz reviewed (commented):
I'm late to the party, so don't know what was said in the welcome speech..

som-snytt commented:
@lrytz squashed, rebased, unwound experiment using replace (I haven't tested performance yet), added pre-check for whether there are escapes to expand.

Probably the ticket doesn't require extreme performance.

Using string concat (stradd) by default assumes that, while we must be JDK 8 compat, hopefully it's using modern intrinsic for adding quote chars, which I assume to be array copy of underlying chars. However, also untested.

som-snytt commented:
```
[info] Benchmark                   Mode  Cnt      Score     Error  Units
[info] ConstantBenchmark.measure   avgt   40  14007.372 ± 176.270  ns/op
[info] ConstantBenchmark.measure0  avgt   40  32518.810 ± 318.486  ns/op
```
I tested by moving the `SymbolTable4Testing` from junit to testkit. I'll PR that in case of utility.

lrytz commented:
Thank you for the benchmark!

> is there no arraycopy if exactly sized?

There must be a copy, the StringBuilder can continue to be used. Unless they do something fancy, like a dirty bit in the string builder and lazy array copy, but I didn't see anything like that.
